File: examples/i-inverse-kinematics.py
from __future__ import print_function

import numpy as np
import pinocchio

model = pinocchio.buildSampleModelManipulator()
data  = model.createData()

JOINT_ID = 6
xdes     = np.matrix([ 0.5,-0.5,0.5]).T

q        = pinocchio.neutral(model)
eps      = 1e-4
IT_MAX   = 1000
DT       = 1e-1

i=0
while True:
    pinocchio.forwardKinematics(model,data,q)
    x   = data.oMi[JOINT_ID].translation
    R   = data.oMi[JOINT_ID].rotation
    err = R.T*(x-xdes)
    if np.linalg.norm(err) < eps:
        print("Convergence achieved!")
        break
    if i >= IT_MAX:
        print("\nWarning: the iterative algorithm has not reached convergence to the desired precision")
        break
    J   = pinocchio.jointJacobian(model,data,q,JOINT_ID)[:3,:]
    v   = - np.linalg.pinv(J)*err
    q   = pinocchio.integrate(model,q,v*DT)
    if not i % 10:        print('error = %s' % err.T)
    i += 1

print('\nresult: %s' % q.flatten().tolist())
print('\nfinal error: %s' % err.T)

File: examples/i-inverse-kinematics.cpp
#include "pinocchio/parsers/sample-models.hpp"
#include "pinocchio/algorithm/kinematics.hpp"
#include "pinocchio/algorithm/jacobian.hpp"
#include "pinocchio/algorithm/joint-configuration.hpp"

int main(int /* argc */, char ** /* argv */)
{
  pinocchio::Model model;
  pinocchio::buildModels::manipulator(model);
  pinocchio::Data data(model);

  const int    JOINT_ID = 6;
  Eigen::Vector3d xdes;        xdes << 0.5, -0.5, 0.5;

  Eigen::VectorXd q     = pinocchio::neutral(model);
  const double eps      = 1e-4;
  const int IT_MAX      = 1000;
  const double DT       = 1e-1;

  pinocchio::Data::Matrix6x J(6,model.nv); J.setZero();
  unsigned int svdOptions = Eigen::ComputeThinU | Eigen::ComputeThinV;
  Eigen::BDCSVD<pinocchio::Data::Matrix3x> svd(3,model.nv,svdOptions);

  Eigen::Vector3d err;
  for (int i=0;;i++)
  {
    pinocchio::forwardKinematics(model,data,q);
    const Eigen::Vector3d & x   = data.oMi[JOINT_ID].translation();
    const Eigen::Matrix3d & R   = data.oMi[JOINT_ID].rotation();
    err = R.transpose()*(x-xdes);
    if(err.norm() < eps)
    {
      std::cout << "Convergence achieved!" << std::endl;
      break;
    }
    if (i >= IT_MAX)
    {
      std::cout << "\nWarning: the iterative algorithm has not reached convergence to the desired precision" << std::endl;
      break;
    }
    pinocchio::jointJacobian(model,data,q,JOINT_ID,J);
    const Eigen::VectorXd v     = - svd.compute(J.topRows<3>()).solve(err);
    q = pinocchio::integrate(model,q,v*DT);
    if(!(i%10)) std::cout << "error = " << err.transpose() << std::endl;
  }

  std::cout << "\nresult: " << q.transpose() << std::endl;
  std::cout << "\nfinal error: " << err.transpose() << std::endl;
}
